File: airplane_mode/airplane_mode/report/revenue_by_airline/revenue_by_airline.py
# Copyright (c) 2025, deepak and contributors
# For license information, please see license.txt

import frappe
from frappe.query_builder import DocType, functions as fn, Order

def execute(filters=None):
    columns = [
        {"label": "Airline", "fieldname": "airline", "fieldtype": "Link", "options": "Airline", "width": 200},
        {"label": "Total Revenue", "fieldname": "total_revenue", "fieldtype": "Currency", "width": 150},
    ]

    Ticket = DocType("Airplane Ticket")
    Airline = DocType("Airline")

    query = (
        frappe.qb.from_(Ticket)
        .join(Airline)
        .on(Ticket.airline == Airline.name)
        .select(
            Ticket.airline,
            fn.Sum(Ticket.total_amount).as_("total_revenue"),
        )
        .groupby(Ticket.airline)
        .orderby(fn.Sum(Ticket.total_amount), order=Order.desc)
    )

    data = query.run(as_dict=True)

    # 🚀 Total Revenue
    total_revenue = sum(d["total_revenue"] for d in data)

    # ✅ Donut Chart
    chart = {
        "data": {
            "labels": [d["airline"] for d in data],
            "datasets": [{"values": [d["total_revenue"] for d in data]}],
        },
        "type": "donut",
        "height": 300,
    }

    # ✅ Report Summary
    summary = [
    {
        "label": "Total Revenue",
        "value": total_revenue,
        "indicator": "green",
        "datatype": "Currency",     # ✅ tells frappe it's money
        "currency": "INR",          # ✅ INR symbol will show
        "background-color": "gray-900"    # ✅ light gray background
    },
    ]


    # # ✅ Add Total Row
    # data.append({"airline": "Total", "total_revenue": total_revenue})

    return columns, data, None, chart, summary

File: airplane_mode/airplane_mode/report/revenue_by_airline/revenue_by_airline.js
// Copyright (c) 2025, deepak and contributors
// For license information, please see license.txt

frappe.pages['revenue-dashboard'].on_page_load = function(wrapper) {
    let page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Revenue Dashboard',
        single_column: true
    });

    // Create chart container
    $(frappe.render_template("<div id='revenue_chart'></div>", {})).appendTo(page.body);

    // Fetch revenue data via frappe.call
    frappe.call({
        method: "frappe.client.get_list",
        args: {
            doctype: "Airplane Ticket",
            fields: ["flight", "total_amount"],
            filters: { status: "Booked" }
        },
        callback: function(r) {
            if (!r.message) return;

            // Group revenue by flight
            let revenueMap = {};
            r.message.forEach(row => {
                if (!revenueMap[row.flight]) {
                    revenueMap[row.flight] = 0;
                }
                revenueMap[row.flight] += row.total_amount;
            });

            // Prepare chart data
            let labels = Object.keys(revenueMap);
            let values = Object.values(revenueMap);

            // Render chart
            new frappe.Chart("#revenue_chart", {
                data: {
                    labels: labels,
                    datasets: [
                        {
                            name: "Total Revenue",
                            values: values
                        }
                    ]
                },
                type: "donut", // or "bar", "line"
                height: 300,
                colors: ["#7cd6fd", "#743ee2", "#ffa3ef", "#ffd700"]
            });
        }
    });
};
